src/rate_limiter.rs

```diff
@@ -813,4 +812,32 @@ mod tests {
             _ => panic!("unexpected result"),
         }
     }
+
+    /// Checks that the rate-limiter's `overcharged` container can't be over-filled.
+    ///
+    /// Keeps trying to add GET requests for `ImmutableData` with a short delay between each
+    /// attempt. Most will fail, but this should ensure the `RateLimiter` always has an entry for
+    /// this client in its `used` container (in case we go back to using the absence of a client as
+    /// a trigger to purge their overcharged entries).
+    ///
+    /// After `OVERCHARGED_TIMEOUT_SECS` plus one minute has elapsed, there should not be an
+    /// excessive number of entries in the `overcharged` container.
+    #[test]
+    fn overcharged_limit() {
+        let mut rate_limiter = RateLimiter::new(false);
+        let client = IpAddr::from([0, 0, 0, 0]);
+        let wait_millis = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 100 / RATE as u64;
+        let max_overcharged_entries = OVERCHARGED_TIMEOUT_SECS * RATE as u64 /
+            MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+        let finish_time = FakeClock::now() + Duration::from_secs(OVERCHARGED_TIMEOUT_SECS + 60);
+        while FakeClock::now() < finish_time {
+            let name = XorName([0; XOR_NAME_LEN]);
+            let msg_id = MessageId::new();
+            let request = UserMessage::Request(Request::GetIData { name, msg_id });
+            let request_parts = unwrap!(request.to_parts(0));
+            let _ = add_user_msg_part(&mut rate_limiter, &client, &request_parts[0]);
+            FakeClock::advance_time(wait_millis);
+        }
+        assert!((rate_limiter.overcharged.len() as u64) <= max_overcharged_entries);
```

Comment:
I think `OVERCHARGED_TIMEOUT_SECS / wait_secs` is too large, because as Fraser said an entry will really only be added every `W = 10 * wait_secs`. However, even with that correction, I think we need to round up, because you can cram e.g. 5 events separated by waits of 10 seconds into a period that is a tiny bit longer than 40 seconds. I did another crummy drawing:

![image](https://sketch.io/render/sk-51f891ad81b78d56209b176e0581c929.jpeg)

Alternatively, we can round up Fraser's original calculation (the test failed and was off-by-one when I changed the `OVERCHARGED_TIMEOUT_SECS` to 300).